src/extensions/default/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/PackageUpdaterStep.cs

```diff
@@ -55,126 +49,183 @@ public class PackageUpdaterStep : UpgradeStep
         };
 
         public PackageUpdaterStep(
-            IPackageRestorer packageRestorer,
-            IEnumerable<IDependencyAnalyzer> packageAnalyzers,
-            IDependencyAnalyzerRunner packageAnalyzer,
+            IPackageRestorer restorer,
+            IEnumerable<IDependencyAnalyzer> dependencyAnalyzers,
             ILogger<PackageUpdaterStep> logger)
             : base(logger)
         {
-            _packageRestorer = packageRestorer ?? throw new ArgumentNullException(nameof(packageRestorer));
-            _packageAnalyzers = packageAnalyzers ?? throw new ArgumentNullException(nameof(packageAnalyzers));
-            _packageAnalyzer = packageAnalyzer ?? throw new ArgumentNullException(nameof(packageAnalyzer));
-            _analysisState = null;
+            _subSteps = dependencyAnalyzers.OrderyByPrecedence().Select(d => new DependencyAnalyzerStep(this, restorer, d, logger)).ToList();
         }
 
         protected override Task<bool> IsApplicableImplAsync(IUpgradeContext context, CancellationToken token) => Task.FromResult(context?.CurrentProject is not null);
 
-        protected override async Task<UpgradeStepInitializeResult> InitializeImplAsync(IUpgradeContext context, CancellationToken token)
+        protected override Task<UpgradeStepInitializeResult> InitializeImplAsync(IUpgradeContext context, CancellationToken token)
         {
-            if (context is null)
+            if (_subSteps.All(d => d.IsDone))
```

Comment:
I think I remember why parent steps often initialize their children. The UpgradeManager will initialize steps (including sub-steps) as it gets to them, but if we wait for that then the steps won't be updated when we do an `IsDone` check in the parent's `InitializeMethod`. So, I suspect that this check isn't going to work the way you expect because the children will all be uninitialized and return false for `IsDone` even if there's no action necessary for them. We'll probably need to initialize the substeps here before doing this check.

Reply:
I saw that exact behavior initially. I then added a check in the AutoApplyStep to also check the parent's children after initializing itself. Now, it will provide the expected behavior, but will still lazily initialize when the UpgradeStep gets to them and then just complete the parent if everything for the parent is now done after initializing.